packages/dapi/test/unit/grpcServer/handlers/blockheaders-stream/subscribeToBlockHeadersWithChainLocksHandlerFactory.spec.js

```diff
@@ -223,4 +229,260 @@ describe('subscribeToBlockHeadersWithChainLocksHandlerFactory', () => {
+    const blockHash = Buffer.from('00000bafbc94add76cb75e2ec92894837288a481e5c005f6563d91623bf8bc2c', 'hex');
+
+    const fakeBlockHeaderHex = '00000020272e374a06c87a0ce0e6ee1a0754c98b9ec2493e7c0ac7ba41a0730000000000568b3c4156090db4d8db5447762e95dd1d4c921c96801a9086720ded85266325916cc05caa94001c5caf3595'; // eslint-disable-line
+    const differentFakeBlockHeaderHex = '000000202be60663802ead0740cb6d6e49ee7824481280f03c71369eb90f7b00000000006abd277facc8cf02886d88662dbcc2adb6d8de7a491915e74bed4d835656a4f1f26dc05ced93001ccf81cabc'; // eslint-disable-line
+
+    let request = new BlockHeadersWithChainLocksRequest();
+
+    beforeEach(() => {
+      request.setFromBlockHash(blockHash);
+      request.setCount(0);
+      request = BlockHeadersWithChainLocksRequest.deserializeBinary(request.serializeBinary());
+
+      call = new GrpcCallMock(sinon, request);
+
+      getHistoricalBlockHeadersIteratorMock = getHistoricalBlockHeadersIteratorFactory(coreAPIMock);
+
+      // eslint-disable-next-line
+      subscribeToBlockHeadersWithChainLocksHandler = subscribeToBlockHeadersWithChainLocksHandlerFactory(
+        getHistoricalBlockHeadersIteratorMock,
+        coreAPIMock,
+        zmqClientMock,
+        subscribeToNewBlockHeadersMock,
+      );
+
+      cache.set(1, undefined);
+      cache.set(2, undefined);
+      cache.set(3, undefined);
+      cache.set(4, undefined);
+      cache.set(5, undefined);
+
+      cacheSpy.get.resetHistory();
+      cacheSpy.set.resetHistory();
+    });
+
+    // the case where we request for N blocks, and theres nothing at all in the cache
+    it('should call for rpc when nothing in the cache', async () => {
+      coreAPIMock.getBlockStats.resolves({ height: 1 });
+      coreAPIMock.getBestBlockHeight.resolves(3);
+      coreAPIMock.getBlockHash.resolves(blockHash.toString('hex'));
+      coreAPIMock.getBlockHeaders.resolves([fakeBlockHeaderHex, fakeBlockHeaderHex,
+        fakeBlockHeaderHex]);
+
+      await subscribeToBlockHeadersWithChainLocksHandler(call);
+
+      expect(coreAPIMock.getBlockHash).to.be.calledOnceWithExactly(1);
+      expect(coreAPIMock.getBlockHeaders).to.be.calledOnceWithExactly(blockHash
+        .toString('hex'), 3);
+
+      expect(cacheSpy.get.callCount).to.be.equal(3);
+      expect(cacheSpy.set.callCount).to.be.equal(3);
+
+      expect(cacheSpy.get).to.always.returned(undefined);
+
+      expect(cacheSpy.set).to.be.calledWith(1, fakeBlockHeaderHex);
+      expect(cacheSpy.set).to.be.calledWith(2, fakeBlockHeaderHex);
+      expect(cacheSpy.set).to.be.calledWith(3, fakeBlockHeaderHex);
+    });
+
+    // the case when we request for cached blocks (all N block are cached)
+    it('should use cache and do not call for blockHeaders', async () => {
+      coreAPIMock.getBlockStats.resolves({ height: 1 });
+      coreAPIMock.getBestBlockHeight.resolves(3);
+      coreAPIMock.getBlockHash.resolves(blockHash.toString('hex'));
+
+      cache.set(1, fakeBlockHeaderHex);
```

Comment:
I suggest using different fake headers in the cache to be closer to the real situation